// ravendb/dist/Http/ServerNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerNode = void 0;
const UriUtility_1 = require("./UriUtility");
class ServerNode {
    constructor(opts) {
        this.clusterTag = null;
        this._lastServerVersionCheck = 0;
        if (opts) {
            this.database = opts.database;
            this.url = opts.url;
            this.clusterTag = opts.clusterTag;
        }
    }
    shouldUpdateServerVersion() {
        if (!this._lastServerVersion || this._lastServerVersionCheck > 100) {
            return true;
        }
        this._lastServerVersionCheck++;
        return false;
    }
    updateServerVersion(serverVersion) {
        this._lastServerVersion = serverVersion;
        this._lastServerVersionCheck = 0;
        this.supportsAtomicClusterWrites = false;
        if (serverVersion) {
            const tokens = serverVersion.split(".");
            try {
                const major = Number.parseInt(tokens[0], 10);
                const minor = Number.parseInt(tokens[1], 10);
                if (major > 5 || (major === 5 && minor >= 2)) {
                    this.supportsAtomicClusterWrites = true;
                }
            }
            catch {
            }
        }
    }
    discardServerVersion() {
        this._lastServerVersion = null;
        this._lastServerVersionCheck = 0;
    }
    static createFrom(topology) {
        const nodes = [];
        if (!topology) {
            return nodes;
        }
        for (const node of Object.keys(topology.members)) {
            const member = topology.members[node];
            nodes.push(new ServerNode({
                url: member,
                clusterTag: node,
                serverRole: "Member"
            }));
        }
        for (const node of Object.keys(topology.watchers)) {
            const watcher = topology.watchers[node];
            nodes.push(new ServerNode({
                url: watcher,
                clusterTag: node,
                serverRole: "Member"
            }));
        }
        return nodes;
    }
    get lastServerVersion() {
        return this._lastServerVersion;
    }
    get isSecure() {
        return UriUtility_1.UriUtility.isSecure(this.url);
    }
    fromJson(json) {
        const from = json;
        this.url = from.Url;
        this.database = from.Database || null;
        this.clusterTag = from.ClusterTag || null;
    }
    static fromJson(json) {
        const node = new ServerNode({
            database: "",
            url: ""
        });
        node.fromJson(json);
        return node;
    }
}
exports.ServerNode = ServerNode;
